### DISCUSSION TITLE
How can I write a customized payload and add into MSF?

### DISCUSSION THREAD

**Discussion by ChinaBluecat** [rapid7/metasploit-framework#15837] [Q&A]
How can I write a customized payload and add into MSF?

I want to use some different payload for my experiment, where can I find document about how payload load into MSF and how to write and install a customized payload. :D

**Comment by bcoles** [comment]
Modules are automatically loaded from the `./modules/payloads` directory (and `~/.msf4/modules/payloads/`).

* https://github.com/rapid7/metasploit-framework/tree/master/modules/payloads

> and how to write and install a customized payload

That depends largely on the type of payload you want to write.

If you're looking to deploy an existing compiled `.exe` file you don't need to write a Metasploit payload. You can instead use the existing [`windows/download_exec`](https://github.com/rapid7/metasploit-framework/blob/master/modules/payloads/singles/windows/download_exec.rb) payload to download and execute it.

**Reply by ChinaBluecat** [reply]
My experiment's arch is windows, but unfortunately, I want to write a new payload to make the connection with the exploit(like a new protocol). I had read some code in ```/modules/payloads```, and it seems that they have some relation with ```/lib/msf/core/payload```, in where some of the binary code are compiled.

**Reply by bcoles** [reply]
> I had read some code in `/modules/payloads`, and it seems that they have some relation with `/lib/msf/core/payload`, in where some of the binary code are compiled.

The payloads use library code from [`lib/msf/core/payload`](https://github.com/rapid7/metasploit-framework/tree/master/lib/msf/core/payload) and the https://github.com/rapid7/metasploit-payloads/ repository. Some also use methods from [Rex](https://github.com/rapid7/rex) libraries.

**Reply by ChinaBluecat** [reply]
Thanks. I should read them a little more. :D
